lib/phoenix/socket/server.ex

```diff
@@ -0,0 +1,160 @@
+defmodule Phoenix.Socket.Supervisor do
+  @moduledoc false
+  use Supervisor
+  alias Phoenix.Socket
+
+  def start_link do
+    Supervisor.start_link(__MODULE__, :ok, name: __MODULE__)
+  end
+
+  def start_child(%Socket{} = socket) do
+    Supervisor.start_child(__MODULE__, [socket])
+  end
+
+  def terminate_child(child) do
+    Supervisor.terminate_child(__MODULE__, child)
+  end
+
+  def init(:ok) do
+    children = [
+      worker(Phoenix.Socket.Server, [], restart: :temporary)
+    ]
+    supervise(children, strategy: :simple_one_for_one)
+  end
+end
+
+
+defmodule Phoenix.Socket.Server do
+  @moduledoc """
+  Defines a server for socket operations. Delegates to the actual Transport.
+  """
+  use GenServer
+  require Logger
+  alias Phoenix.Socket
+  alias Phoenix.Socket.Message
+  alias Phoenix.PubSub
+
+  # External API
+
+  def start_link(%Socket{} = socket) do
+    GenServer.start_link(__MODULE__, socket)
+  end
+
+  def dispatch_join(server, topic, message) do
+    GenServer.call(server, {:dispatch_join, topic, message})
+  end
+
+  def dispatch_in(server, topic, event, message) do
+    GenServer.call(server, {:dispatch_in, topic, event, message})
+  end
+
+  def dispatch_out(server, event, message) do
+    GenServer.call(server, {:dispatch_out, event, message})
+  end
+
+  def dispatch_leave(server, message) do
+    GenServer.call(server, {:dispatch_leave, :ignore_topic, message})
+  end
+  def dispatch_leave(server, topic, message) do
+    GenServer.call(server, {:dispatch_leave, topic, message})
+  end
+
+  def do_join(server) do
+    GenServer.call(server, :do_join)
+  end
+
+  def do_leave(server) do
+    GenServer.call(server, :do_leave)
+  end
+
+  def authorized?(server, topic) do
+    GenServer.call(server, {:authorized, topic})
+  end
+
+  def dispatch_reply(server, %Message{} = message) do
+    GenServer.cast(server, {:dispatch_reply, message})
+  end
+
+  # Server internal
+
+  def init(%Socket{} = socket) do
+    {:ok, socket}
+  end
+
+  def handle_call({:dispatch_join, topic, message}, _from, socket) do
+    case socket.router.channel_for_topic(topic, socket.transport) do
```

Comment:
I think this should be in the transport. The channel should not care about the router or transport it came from. :)